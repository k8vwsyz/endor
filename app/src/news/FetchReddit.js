const fetchNews = async () => {
  // "Hot" posts all time with the Environment flair
  const url =
    "https://www.reddit.com/r/science/search.json?q=flair:Environment&restrict_sr=on&sort=hot&t=all&limit=1000";

  const news = await fetch(url).then((resp) =>
    resp.json().then((posts) => mapPosts(posts.data.children))
  );

  return news;
};

const mapPosts = (posts) => {
  return posts.map((post) => ({
    title: post.data.title,
    thumbnail: post.data.thumbnail,
    url: post.data.url,
    created: post.data.created_utc,
  }));
};

let filterNews = news => {
  const keyWords = ["iceberg", "antarctic", "ocean", "lagoon", "waterhole", "marine", "fish", "algae", "algal Blooms", "water", "cryptosporidium", "effluent", "oil spill", "river basin", "sewage", "river", "acequia", "alluvium", "aquaculture", "aqueduct", "aquifer", "artificial recharge", "base flow", "bay", "bedrock", "brook", "condensation", "consumptive use", "conveyance loss", "creek", "cubic feet per second", "cfs", "desalination", "dew", "drainage basin", "drawdown", "drought", "drip irrigation", "effluent", "erosion", "estuary", "evaporation", "evapotranspiration", "flood", "floodway", "flowing", "well", "spring", "freshwater", "fresh water", "gage height", "gaging station", "geyser", "giardiasis", "glacier", "greywater", "groundwater", "headwater", "hydrologic cycle", "ice", "impermeable layer", "infiltration", "irrigation", "lake", "leaching", "lentic waters", "levee", "waters", "marsh", "gallons", "osmosis", "outfall", "peak flow", "percolation", "permeability", "porosity", "precipitation", "wastewater", "reach", "recharge", "reservoir", "flow", "runoff", "sea", "sediment", "sedimentary", "sedimentation", "seepage", "septic tank", "pond", "sewage", "sinkhole", "snow", "sprain", "irrigation", "steam", "stage", "sewer", "stream", "streamflow", "subsidence", "surface tension", "transpiration", "tributary", "turbidity", "unsaturated zone", "vapor", "watershed", "xeriscaping"]

  return news.filter(n => keyWords.some(k => n.title.includes(k)));
}

// Look for url dupes and remove them
let removeDupes = news => news.filter((news1, i, arr) => i === arr.findIndex(news2 => news2.url === news1.url))

export { fetchNews, removeDupes, filterNews };

