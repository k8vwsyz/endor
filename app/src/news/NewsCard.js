import Box from "@material-ui/core/Box";
import Card from "@material-ui/core/Card";
import { default as CardActionArea } from "@material-ui/core/CardActionArea";
import CardMedia from "@material-ui/core/CardMedia";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import React from "react";

const useStyles = makeStyles(theme => ({
  root: {
    width: 540,
    height: "fit-content",
    minWidth: 130,
    marginTop: 12,
    marginBottom: 12,
  },
  box: {
    paddingBottom: 12,
    display: "flex",
    justifyContent: "space-between",
  },
  cover: {
    height: "5em",
    width: "5em",
    borderRadius: "5%",
  },
  website: {
    paddingLeft: "5px",
    textTransform: "capitalize",
    fontWeight: "bold",
  },
}));

export default function NewsCard(props) {
  const classes = useStyles();
  const { url, thumbnail, summary, created, title, favicon } = props;
  const urlRegex = "^(?:https?://)?(?:[^@\n]+@)?(?:www.)?([^:/\n?]+)";

  const titleDisplayed = title
    ? title
    : url
    ? url.match(urlRegex)[1].split(".")[0]
    : "";
  const faviconDisplayed = favicon
    ? favicon
    : url
    ? "https://www.google.com/s2/favicons?domain=" + url.match(urlRegex)[0]
    : "";
  return (
    <Card className={classes.root} raised={true}>
      <CardActionArea
        onClick={() => url === "/news" ? window.open(url, "_self") : window.open(url) }
        style={{ width: "100%", height: "100%" }}
      >
        <Box style={{ padding: 12, height: "100%" }}>
          <Box className={classes.box}>
            <Box>
              <Box style={{ display: "flex", alignItems: "center" }}>
                <img alt="news item thumbnail" src={faviconDisplayed} />
                <Typography variant="h6" className={classes.website}>
                  {titleDisplayed}
                </Typography>
              </Box>
              <Typography variant="caption">
                {convertUnixTimestamp(created)}
              </Typography>
            </Box>
            <CardMedia
              data-testid="cardmedia"
              className={classes.cover}
              component="img"
              height="140"
              image={
                thumbnail === "default" || thumbnail === "" || !thumbnail
                  ? "/pictures/defaultWaterPicture.jpg"
                  : thumbnail
              }
            />
          </Box>
          <Typography variant="body2">{summary}</Typography>
        </Box>
      </CardActionArea>
    </Card>
  );
}
function convertUnixTimestamp(unixTimestamp) {
  const a = new Date(unixTimestamp * 1000);
  const months = [ "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec", ];
  const year = a.getFullYear();
  const month = months[a.getMonth()];
  const date = a.getDate();
  return date + " " + month + " " + year;
}
