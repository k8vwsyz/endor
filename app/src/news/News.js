import { makeStyles } from "@material-ui/core/styles";
import React, { useEffect, useState } from "react";
import LoadingScreen from "../loadingscreen/Loadingscreen";
import { fetchNews, filterNews, removeDupes } from "./FetchReddit";
import NewsCard from "./NewsCard";

const useStyles = makeStyles({
  root: {
    display: "flex",
    flexDirection: "row",
    flexWrap: "wrap",
    justifyContent: "space-evenly",
  },
});

export default function News() {
  const [news, setNews] = useState([]);
  const [loading, setLoading] = useState(true);
  const classes = useStyles();

  // Only fetch the news on component mount
  useEffect(() => {
    const newsEffect = async () => {
      const newsItems = await fetchNews();
      const filteredNews = filterNews(newsItems);
      const dupesRemoved = removeDupes(filteredNews);
      // dupesRemoved.sort((n1, n2) => n2.title.length - n1.title.length);
      dupesRemoved.sort((n1, n2) => n2.created - n1.created);
      // setTimeout(() => { //to make the loading screen appear for 1,5 sec
      setNews(dupesRemoved);
      setLoading(false);
      // }, 1500)           // I belong to the comment over me.
    };
    newsEffect();
  }, []);

  if (!loading)
    return (
      <div className={classes.root} data-testid="news">
        {news.map((newsItem, i) => (
          <NewsCard
            key={i}
            summary={newsItem.title}
            thumbnail={newsItem.thumbnail}
            url={newsItem.url}
            created={newsItem.created}
          />
        ))}
      </div>
    );
  else return <LoadingScreen />;
}
