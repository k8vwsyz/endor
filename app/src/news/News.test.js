import { waitFor } from "@testing-library/dom";
import { act, cleanup, render, screen } from "@testing-library/react";
import React from "react";
import defaultWaterPicture from "../../public/pictures/defaultWaterPicture.jpg";
import * as reddit from "./FetchReddit";
import News from "./News";
import NewsCard from "./NewsCard";

afterEach(cleanup);

reddit.fetchNews = jest.fn(() => [
  {
    title: "Title!",
    thumbnail: defaultWaterPicture,
    url: "https://material-ui.com/components/cards/",
  },
  {
    title: "Title 2!",
    thumbnail:
      "https://b.thumbs.redditmedia.com/M56Fqi-q0KlTlIlAFiHUG3fYaU1nGrWFMVnXvSy1Msc.jpg",
    url: "https://material-ui.com/components/cards/",
  },
  {
    title: "water",
    thumbnail: defaultWaterPicture,
    url: "https://material-ui.com/components/cards/",
  },
  {
    title: "water",
    thumbnail: defaultWaterPicture,
    url: "https://material-ui.com/components/cards/",
  },
]);
describe("Snapshot tests", () => {
  it("test styles", async () => {
    const { getByTestId, asFragment } = render(<News />);
    await act(async () => waitFor(() => getByTestId("news")));
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Title tests", () => {
  it("test title 1", async () => {
    await act(async () => render(<News />));
    expect(screen.getByText("water")).toBeInTheDocument();
  });
});

describe("Thumbnail tests", () => {
  it("displays the correct external picture", () => {
    const { container } = render(
      <NewsCard thumbnail="https://b.thumbs.redditmedia.com/M56Fqi-q0KlTlIlAFiHUG3fYaU1nGrWFMVnXvSy1Msc.jpg" />
    );
    expect(container.querySelector("button > div > div > img").src).toBe(
      "https://b.thumbs.redditmedia.com/M56Fqi-q0KlTlIlAFiHUG3fYaU1nGrWFMVnXvSy1Msc.jpg"
    );
  });

  it('displays the default picture when thumbnail is "default"', () => {
    const { container } = render(<NewsCard thumbnail="default" />);
    expect(container.querySelector("button > div > div > img").src).toBe(
      "http://localhost/pictures/defaultWaterPicture.jpg"
    );
  });

  it("displays the default picture when empty thumbnail", () => {
    const { container } = render(<NewsCard thumbnail="" />);
    expect(container.querySelector("button > div > div > img").src).toBe(
      "http://localhost/pictures/defaultWaterPicture.jpg"
    );
  });

  it("displays the default picture when no thumbnail", () => {
    const { container } = render(<NewsCard />);
    expect(container.querySelector("button > div > div > img").src).toBe(
      "http://localhost/pictures/defaultWaterPicture.jpg"
    );
  });
});

describe("Filter & removeDupes", () => {
  it("removes irrelevant news", async () => {
    await act(async () => render(<News />));
    expect(screen.getByTestId("news").childElementCount).toBe(1);
  });
});
