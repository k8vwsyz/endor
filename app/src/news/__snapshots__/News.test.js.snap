// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests test styles 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
    data-testid="news"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-6 MuiPaper-elevation8 MuiPaper-rounded"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root"
        style="width: 100%; height: 100%;"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiBox-root MuiBox-root-10"
          style="padding: 12px; height: 100%;"
        >
          <div
            class="MuiBox-root MuiBox-root-11 makeStyles-box-7"
          >
            <div
              class="MuiBox-root MuiBox-root-12"
            >
              <div
                class="MuiBox-root MuiBox-root-13"
                style="display: flex; align-items: center;"
              >
                <img
                  alt="news item thumbnail"
                  src="https://www.google.com/s2/favicons?domain=https://material-ui.com"
                />
                <h6
                  class="MuiTypography-root makeStyles-website-9 MuiTypography-h6"
                >
                  material-ui
                </h6>
              </div>
              <span
                class="MuiTypography-root MuiTypography-caption"
              >
                NaN undefined NaN
              </span>
            </div>
            <img
              class="MuiCardMedia-root makeStyles-cover-8 MuiCardMedia-media MuiCardMedia-img"
              data-testid="cardmedia"
              height="140"
              src="defaultWaterPicture.jpg"
            />
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2"
          >
            water
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
